File: fconfworkshop/src/CourseExamples/Module4/AgentStocks/AgentStocks/ViewModel/MainWindowViewModel.fs
namespace FSharpWpfMvvmTemplate.ViewModel

open System
open System.Collections.ObjectModel
open System.Windows
open System.Windows.Data
open System.Windows.Input
open System.ComponentModel
open System.Collections.Generic
open OxyPlot
open OxyPlot.Axes
open OxyPlot.Series
open Messages


type MainWindowViewModel() =
    inherit ViewModelBase()

    let mutable plotModel =
        let p = PlotModel()
        p.LegendTitle <- "Legend"
        p.LegendOrientation <- LegendOrientation.Horizontal
        p.LegendPlacement <- LegendPlacement.Outside
        p.LegendPosition <- LegendPosition.TopRight
        p.LegendBackground <- OxyColor.FromAColor(200uy, OxyColors.White)
        p.LegendBorder <- OxyColors.Black
        p

    let mutable stockButtonViewModels = Dictionary<string, StockToggleButtonViewModel>()

    let chartingActorRef = Agents.lineChartingAgent(plotModel)
    let stocksCoordinatorActorRef = Agents.stocksCoordinatorAgent(chartingActorRef)
    
    do
        let stockDateTimeAxis = new DateTimeAxis()
        stockDateTimeAxis.Position <- AxisPosition.Bottom
        stockDateTimeAxis.MajorGridlineStyle <- LineStyle.Solid
        stockDateTimeAxis.MinorGridlineStyle <- LineStyle.Dot
        stockDateTimeAxis.Title <- "Date"
        stockDateTimeAxis.StringFormat <- "HH:mm:ss"
        plotModel.Axes.Add(stockDateTimeAxis)

        let stockPriceAxis = new LinearAxis()
        stockPriceAxis.Minimum <- 0.
        stockPriceAxis.MajorGridlineStyle <- LineStyle.Solid
        stockPriceAxis.MinorGridlineStyle <- LineStyle.Dot
        stockPriceAxis.Title <- "Price"
        plotModel.Axes.Add(stockPriceAxis)

        for stock in ["MSFT"; "APPL"; "GOOG"] do
            let  s = StockToggleButtonViewModel(stock, stocksCoordinatorActorRef)
            stockButtonViewModels.Add(stock, s)
        
    member self.StockButtonViewModels 
        with get() = stockButtonViewModels
        and set value = stockButtonViewModels <- value

    member self.PlotModel
        with get() = plotModel
        and set value = 
            plotModel <- value
            base.OnPropertyChanged("PlotModel")

        

File: fconfworkshop/src/reactivestock/ReactiveStock/ViewModel/MainWindowViewModel.cs
using System.Collections.Generic;
using GalaSoft.MvvmLight;
using OxyPlot;
using OxyPlot.Axes;
using ReactiveStock.ActorModel;
using ReactiveStock.ActorModel.Actors;
using ReactiveStock.ActorModel.Actors.UI;
using ReactiveStock.ActorModel.Actors.Core;

namespace ReactiveStock.ViewModel
{
    public class MainWindowViewModel : ViewModelBase
    {
        private IAgent<ChartSeriesMessage> _chartingActorRef;
        private IAgent<StocksCoordinatorMessage> _stocksCoordinatorActorRef;
        private PlotModel _plotModel;

        public Dictionary<string, StockToggleButtonViewModel> StockButtonViewModels { get; set; }

        public PlotModel PlotModel
        {
            get { return _plotModel; }
            set { Set(() => PlotModel, ref _plotModel, value); }
        }

        public MainWindowViewModel()
        {
            SetUpChartModel();

            InitializeActors();

            CreateStockButtonViewModels();
        }


        private void SetUpChartModel()
        {
            _plotModel = new PlotModel
            {
                LegendTitle = "Legend",
                LegendOrientation = LegendOrientation.Horizontal,
                LegendPlacement = LegendPlacement.Outside,
                LegendPosition = LegendPosition.TopRight,
                LegendBackground = OxyColor.FromAColor(200, OxyColors.White),
                LegendBorder = OxyColors.Black
            };


            var stockDateTimeAxis = new DateTimeAxis
            {
                Position = AxisPosition.Bottom,
                MajorGridlineStyle = LineStyle.Solid,
                MinorGridlineStyle = LineStyle.Dot,
                Title = "Date",
                StringFormat = "HH:mm:ss"
            };

            _plotModel.Axes.Add(stockDateTimeAxis);


            var stockPriceAxis = new LinearAxis
            {
                Minimum = 0,
                MajorGridlineStyle = LineStyle.Solid,
                MinorGridlineStyle = LineStyle.Dot,
                Title = "Price"
            };

            _plotModel.Axes.Add(stockPriceAxis);
        }


        private void InitializeActors()
        {
            _chartingActorRef =
                new LineChartingActor(PlotModel).Actor;

            _stocksCoordinatorActorRef =
                new StocksCoordinatorActor(_chartingActorRef).Actor;
        }


        private void CreateStockButtonViewModels()
        {
            StockButtonViewModels = new Dictionary<string, StockToggleButtonViewModel>();

            CreateStockButtonViewModel("AAAA");
            CreateStockButtonViewModel("BBBB");
            CreateStockButtonViewModel("CCCC");
        }

        private void CreateStockButtonViewModel(string stockSymbol)
        {
            var newViewModel = new StockToggleButtonViewModel(_stocksCoordinatorActorRef, stockSymbol);

            StockButtonViewModels.Add(stockSymbol, newViewModel);
        }
    }
}